include/ginkgo/core/reorder/nested_dissection.hpp

```diff
@@ -0,0 +1,148 @@
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GKO_PUBLIC_CORE_REORDER_NESTED_DISSECTION_HPP_
+#define GKO_PUBLIC_CORE_REORDER_NESTED_DISSECTION_HPP_
+
+
+#include <unordered_map>
+
+
+#include <ginkgo/config.hpp>
+
+
+#if GKO_HAVE_METIS
+
+
+#include <memory>
+
+
+#include <ginkgo/core/base/abstract_factory.hpp>
+#include <ginkgo/core/base/array.hpp>
+#include <ginkgo/core/base/dim.hpp>
+#include <ginkgo/core/base/lin_op.hpp>
+#include <ginkgo/core/base/polymorphic_object.hpp>
+#include <ginkgo/core/base/types.hpp>
+#include <ginkgo/core/matrix/csr.hpp>
+#include <ginkgo/core/matrix/permutation.hpp>
+
+
+namespace gko {
+namespace experimental {
+/**
+ * @brief The Reorder namespace.
+ *
+ * @ingroup reorder
+ */
+namespace reorder {
+
+
+/**
+ * Computes a Nested Dissection (ND) reordering of an input matrix using the
+ * METIS library.
+ *
+ * @tparam ValueType  the type used to store values of the system matrix
+ * @tparam IndexType  the type used to store sparsity pattern indices of the
+ *                    system matrix
+ */
+template <typename ValueType, typename IndexType>
+class NestedDissection
+    : public EnablePolymorphicObject<NestedDissection<ValueType, IndexType>,
+                                     LinOpFactory>,
+      public EnablePolymorphicAssignment<
+          NestedDissection<ValueType, IndexType>> {
+public:
+    struct parameters_type;
+    friend class EnablePolymorphicObject<NestedDissection<ValueType, IndexType>,
+                                         LinOpFactory>;
+    friend class enable_parameters_type<parameters_type,
+                                        NestedDissection<ValueType, IndexType>>;
+
+    using value_type = ValueType;
+    using index_type = IndexType;
+    using matrix_type = matrix::Csr<value_type, index_type>;
+    using permutation_type = matrix::Permutation<index_type>;
+
+    struct parameters_type
+        : public enable_parameters_type<
+              parameters_type, NestedDissection<ValueType, IndexType>> {
+        /**
+         * The options to be passed on to METIS, stored as key-value pairs.
+         * Any options that are not set here use their default value.
+         */
+        std::unordered_map<int, int> options;
```

Comment:
I think the unodered_map<int, int> is for METIS only. If the NestedDissection is only available for METIS, I will prefer having METIS in class

Reply:
At the moment it's only METIS, but in the future I would add interfaces to [KaHIP](https://kahip.github.io/) and [Scotch/PT-Scotch](https://www.labri.fr/perso/pelegrin/scotch/)

Reply:
do they also use the unordered_map for their parameters?

Reply:
Not really, KaHIP has only a single `mode` parameter and Scotch has an opaque Strategy pointer. But we can add those when we add support for them. Do you think it should be called `metis_options` instead?

Reply:
maybe keeping it as options to leave possiblity to reuse it